File: android-app/app/src/main/java/com/plantgotchi/app/agent/RuleEngine.kt
package com.plantgotchi.app.agent

import com.plantgotchi.app.model.Phase
import com.plantgotchi.app.model.Plant
import com.plantgotchi.app.model.Recommendation
import com.plantgotchi.app.model.SensorReading
import java.util.UUID
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * Offline rule engine — runs on every new sensor reading, no network needed.
 * Returns recommendations based on simple threshold checks.
 *
 * Port of `website-astro/src/lib/agents/rules.ts` and `plant-view.ts`.
 */
object RuleEngine {

    data class ResolvedThresholds(
        val tempMin: Double,
        val tempMax: Double,
        val moistureMin: Int,
        val moistureMax: Int,
    )

    fun resolveThresholds(plant: Plant): ResolvedThresholds {
        val phase = Phase.fromValue(plant.currentPhase)

        // Always use plant-level moisture
        val moistureMin = plant.moistureMin
        val moistureMax = plant.moistureMax

        // Use phase defaults for temp when phase is set
        if (phase != null && phase != Phase.COMPLETE) {
            val defaults = phase.defaults
            if (defaults != null) {
                return ResolvedThresholds(
                    tempMin = defaults.tempMin,
                    tempMax = defaults.tempMax,
                    moistureMin = moistureMin,
                    moistureMax = moistureMax,
                )
            }
        }

        return ResolvedThresholds(
            tempMin = plant.tempMin,
            tempMax = plant.tempMax,
            moistureMin = moistureMin,
            moistureMax = moistureMax,
        )
    }

    fun getTransitionSuggestions(plant: Plant): List<Recommendation> {
        val phase = Phase.fromValue(plant.currentPhase) ?: return emptyList()
        val startedAt = plant.phaseStartedAt ?: return emptyList()

        val daysInPhase = try {
            val start = java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'").apply {
                timeZone = java.util.TimeZone.getTimeZone("UTC")
            }.parse(startedAt)?.time ?: return emptyList()
            ((System.currentTimeMillis() - start) / 86400000).toInt()
        } catch (_: Exception) {
            return emptyList()
        }

        val recs = mutableListOf<Recommendation>()

        if (phase == Phase.VEGETATIVE && daysInPhase >= 42) {
            recs.add(Recommendation(
                id = UUID.randomUUID().toString(),
                plantId = plant.id,
                source = "rules",
                message = "${plant.name} has been in vegetative for $daysInPhase days — consider switching to flower (12/12 light cycle)",
                severity = "info",
            ))
        }

        if (phase == Phase.DRYING && daysInPhase >= 7) {
            recs.add(Recommendation(
                id = UUID.randomUUID().toString(),
                plantId = plant.id,
                source = "rules",
                message = "${plant.name} has been drying for $daysInPhase days — check the stem snap test to see if it's ready for curing",
                severity = "info",
            ))
        }

        return recs
    }

    /**
     * Evaluate a plant against its latest sensor reading and produce recommendations.
     *
     * @param plant The plant with configured thresholds.
     * @param reading The latest sensor reading for that plant.
     * @return A list of recommendations (may be empty if everything is OK).
     */
    fun evaluatePlant(plant: Plant, reading: SensorReading): List<Recommendation> {
        // Skip non-monitored phases
        val phase = Phase.fromValue(plant.currentPhase)
        if (phase != null && !phase.hasMonitoring) return emptyList()

        val thresholds = resolveThresholds(plant)
        val recs = mutableListOf<Recommendation>()

        // Moisture too low
        if (reading.moisture != null && reading.moisture < thresholds.moistureMin) {
            recs.add(
                Recommendation(
                    id = UUID.randomUUID().toString(),
                    plantId = plant.id,
                    source = "rules",
                    message = "${plant.name} needs water! Soil moisture at ${reading.moisture}% (minimum: ${thresholds.moistureMin}%)",
                    severity = if (reading.moisture < 20) "urgent" else "warning",
                )
            )
        }

        // Moisture too high
        if (reading.moisture != null && reading.moisture > thresholds.moistureMax) {
            recs.add(
                Recommendation(
                    id = UUID.randomUUID().toString(),
                    plantId = plant.id,
                    source = "rules",
                    message = "${plant.name} may be overwatered — moisture at ${reading.moisture}% (maximum: ${thresholds.moistureMax}%)",
                    severity = "warning",
                )
            )
        }

        // Temperature too low
        if (reading.temperature != null && reading.temperature < thresholds.tempMin) {
            recs.add(
                Recommendation(
                    id = UUID.randomUUID().toString(),
                    plantId = plant.id,
                    source = "rules",
                    message = "Too cold for ${plant.name}! Temperature at ${reading.temperature}\u00B0C (minimum: ${thresholds.tempMin}\u00B0C)",
                    severity = if (reading.temperature < thresholds.tempMin - 5) "urgent" else "warning",
                )
            )
        }

        // Temperature too high
        if (reading.temperature != null && reading.temperature > thresholds.tempMax) {
            recs.add(
                Recommendation(
                    id = UUID.randomUUID().toString(),
                    plantId = plant.id,
                    source = "rules",
                    message = "Too hot for ${plant.name}! Temperature at ${reading.temperature}\u00B0C (maximum: ${thresholds.tempMax}\u00B0C)",
                    severity = if (reading.temperature > thresholds.tempMax + 5) "urgent" else "warning",
                )
            )
        }

        // Low battery
        if (reading.battery != null && reading.battery < 15) {
            recs.add(
                Recommendation(
                    id = UUID.randomUUID().toString(),
                    plantId = plant.id,
                    source = "rules",
                    message = "Sensor battery low for ${plant.name} (${reading.battery}%) — charge soon",
                    severity = if (reading.battery < 5) "urgent" else "warning",
                )
            )
        }

        // High-light plant not getting enough light — only in growing phases or no phase
        if (reading.light != null && plant.lightPreference == "high" && reading.light < 1000) {
            if (phase == null || phase.isGrowing) {
                recs.add(
                    Recommendation(
                        id = UUID.randomUUID().toString(),
                        plantId = plant.id,
                        source = "rules",
                        message = "${plant.name} prefers bright light but only getting ${reading.light} lux — consider moving to a sunnier spot",
                        severity = "info",
                    )
                )
            }
        }

        // Add transition suggestions
        recs.addAll(getTransitionSuggestions(plant))

        return recs
    }

    // ---- Plant view helpers (ported from plant-view.ts) ----

    /**
     * Returns a human-readable label for a light level in lux.
     */
    fun getLightLabel(light: Int?): String {
        if (light == null) return "unknown"
        return when {
            light < 1000 -> "low"
            light <= 2000 -> "medium"
            else -> "high"
        }
    }

    /**
     * Compute a "health points" score (0-100) based on how close
     * moisture and temperature are to the midpoint of their ideal ranges.
     */
    fun computeHP(
        moisture: Int?,
        temp: Double?,
        moistureMin: Int,
        moistureMax: Int,
        tempMin: Double,
        tempMax: Double,
    ): Int {
        val scores = mutableListOf<Double>()

        if (moisture != null) {
            val mid = (moistureMin + moistureMax) / 2.0
            val half = (moistureMax - moistureMin) / 2.0
            if (half > 0) {
                val score = max(0.0, min(100.0, 100.0 - (abs(moisture - mid) / half) * 100.0))
                scores.add(score)
            } else {
                scores.add(if (moisture.toDouble() == mid) 100.0 else 0.0)
            }
        }

        if (temp != null) {
            val mid = (tempMin + tempMax) / 2.0
            val half = (tempMax - tempMin) / 2.0
            if (half > 0) {
                val score = max(0.0, min(100.0, 100.0 - (abs(temp - mid) / half) * 100.0))
                scores.add(score)
            } else {
                scores.add(if (temp == mid) 100.0 else 0.0)
            }
        }

        if (scores.isEmpty()) return 50
        return (scores.sum() / scores.size).roundToInt()
    }

    /**
     * Compute a plant status based on whether moisture and temperature
     * are within their configured ranges.
     */
    fun computeStatus(
        moisture: Int?,
        temp: Double?,
        moistureMin: Int,
        moistureMax: Int,
        tempMin: Double,
        tempMax: Double,
    ): String {
        if (moisture == null && temp == null) return "unknown"

        if (moisture != null && (moisture < moistureMin || moisture > moistureMax)) {
            return "thirsty"
        }
        if (temp != null && (temp < tempMin || temp > tempMax)) {
            return "thirsty"
        }

        return "happy"
    }
}

File: ios-app/Plantgotchi/Agents/RuleEngine.swift
import Foundation

// MARK: - Plant View Helpers (ported from website-astro/src/lib/plant-view.ts)

/// Status of a plant based on current sensor readings.
enum PlantStatus: String, Equatable {
    case happy
    case thirsty
    case unknown
}

/// Returns a human-readable light level label.
/// Ported from `getLightLabel()` in plant-view.ts.
func getLightLabel(light: Int?) -> String {
    guard let light = light else { return "unknown" }
    if light < 1000 { return "low" }
    if light <= 2000 { return "medium" }
    return "high"
}

/// Compute a health-point score (0-100) based on how close moisture and temperature
/// are to the midpoint of their acceptable range.
/// Ported from `computeHP()` in plant-view.ts.
func computeHP(
    moisture: Int?,
    temp: Double?,
    moistureMin: Int,
    moistureMax: Int,
    tempMin: Double,
    tempMax: Double
) -> Int {
    var scores: [Double] = []

    if let moisture = moisture {
        let mid = Double(moistureMin + moistureMax) / 2.0
        let half = Double(moistureMax - moistureMin) / 2.0
        let score: Double
        if half == 0 {
            score = moisture == Int(mid) ? 100 : 0
        } else {
            score = max(0, min(100, 100 - (abs(Double(moisture) - mid) / half) * 100))
        }
        scores.append(score)
    }

    if let temp = temp {
        let mid = (tempMin + tempMax) / 2.0
        let half = (tempMax - tempMin) / 2.0
        let score: Double
        if half == 0 {
            score = temp == mid ? 100 : 0
        } else {
            score = max(0, min(100, 100 - (abs(temp - mid) / half) * 100))
        }
        scores.append(score)
    }

    if scores.isEmpty { return 50 }
    return Int((scores.reduce(0, +) / Double(scores.count)).rounded())
}

/// Determine the plant's overall status based on sensor readings and thresholds.
/// Ported from `computeStatus()` in plant-view.ts.
func computeStatus(
    moisture: Int?,
    temp: Double?,
    moistureMin: Int,
    moistureMax: Int,
    tempMin: Double,
    tempMax: Double
) -> PlantStatus {
    if moisture == nil && temp == nil { return .unknown }

    if let moisture = moisture, (moisture < moistureMin || moisture > moistureMax) {
        return .thirsty
    }
    if let temp = temp, (temp < tempMin || temp > tempMax) {
        return .thirsty
    }

    return .happy
}

// MARK: - Plant View (composite)

/// A computed view-model for a plant, combining plant data with latest readings.
struct PlantView {
    let id: String
    let name: String
    let species: String
    let emoji: String
    let moisture: Int?
    let temp: Double?
    let light: Int?
    let lightLabel: String
    let lastWatered: String?
    let status: PlantStatus
    let hp: Int
    let moistureMin: Int
    let moistureMax: Int
    let tempMin: Double
    let tempMax: Double
    let phase: String?
}

/// Build a `PlantView` from a plant, its latest reading, and recent care logs.
/// Ported from `toPlantView()` in plant-view.ts.
func toPlantView(
    plant: Plant,
    latestReading: SensorReading?,
    recentCareLogs: [CareLog]
) -> PlantView {
    let moisture = latestReading?.moisture
    let temp = latestReading?.temperature
    let light = latestReading?.light

    let waterLog = recentCareLogs.first { $0.action == "water" }

    return PlantView(
        id: plant.id,
        name: plant.name,
        species: plant.species ?? "",
        emoji: plant.emoji,
        moisture: moisture,
        temp: temp,
        light: light,
        lightLabel: getLightLabel(light: light),
        lastWatered: waterLog?.createdAt,
        status: computeStatus(
            moisture: moisture,
            temp: temp,
            moistureMin: plant.moistureMin,
            moistureMax: plant.moistureMax,
            tempMin: plant.tempMin,
            tempMax: plant.tempMax
        ),
        hp: computeHP(
            moisture: moisture,
            temp: temp,
            moistureMin: plant.moistureMin,
            moistureMax: plant.moistureMax,
            tempMin: plant.tempMin,
            tempMax: plant.tempMax
        ),
        moistureMin: plant.moistureMin,
        moistureMax: plant.moistureMax,
        tempMin: plant.tempMin,
        tempMax: plant.tempMax,
        phase: plant.currentPhase?.rawValue
    )
}

// MARK: - Rule Engine (ported from website-astro/src/lib/agents/rules.ts)

/// Resolved thresholds used by the rule engine.
struct ResolvedThresholds {
    let tempMin: Double
    let tempMax: Double
    let moistureMin: Int
    let moistureMax: Int
}

/// Offline rule engine. Evaluates a plant + reading against threshold rules
/// and returns any generated recommendations. No network required.
enum RuleEngine {

    // MARK: - Threshold Resolution

    /// Returns effective thresholds for a plant, taking its current phase into account.
    /// - If the plant has a phase with monitoring: use phase defaults for temp, plant-level for moisture.
    /// - For flowering phase: if phaseStartedAt is >= 6 weeks ago, use lateFlowerDefaults for temp.
    /// - If no phase (nil): fall back to plant-level thresholds (backward compat).
    static func resolveThresholds(plant: Plant) -> ResolvedThresholds {
        guard let phase = plant.currentPhase else {
            // No phase set — backward compat: use plant-level thresholds
            return ResolvedThresholds(
                tempMin: plant.tempMin,
                tempMax: plant.tempMax,
                moistureMin: plant.moistureMin,
                moistureMax: plant.moistureMax
            )
        }

        guard phase.hasMonitoring else {
            // Non-monitored phases still return plant-level (won't be used since
            // evaluatePlant returns early, but keeps the API consistent)
            return ResolvedThresholds(
                tempMin: plant.tempMin,
                tempMax: plant.tempMax,
                moistureMin: plant.moistureMin,
                moistureMax: plant.moistureMax
            )
        }

        // Determine which phase defaults to use for temperature
        let phaseDefaults: PhaseDefaults
        if phase == .flowering, let startStr = plant.phaseStartedAt {
            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            let startDate = formatter.date(from: startStr)
                ?? ISO8601DateFormatter().date(from: startStr)
            if let start = startDate {
                let sixWeeks = TimeInterval(6 * 7 * 24 * 60 * 60)
                if Date().timeIntervalSince(start) >= sixWeeks {
                    phaseDefaults = Phase.lateFlowerDefaults
                } else {
                    phaseDefaults = phase.defaults
                }
            } else {
                phaseDefaults = phase.defaults
            }
        } else {
            phaseDefaults = phase.defaults
        }

        return ResolvedThresholds(
            tempMin: phaseDefaults.tempMinC,
            tempMax: phaseDefaults.tempMaxC,
            moistureMin: plant.moistureMin,
            moistureMax: plant.moistureMax
        )
    }

    // MARK: - Transition Suggestions

    /// Returns lifecycle transition suggestions based on phase duration and plant type.
    static func transitionSuggestions(plant: Plant) -> [Recommendation] {
        guard let phase = plant.currentPhase,
              let startStr = plant.phaseStartedAt else {
            return []
        }

        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        guard let startDate = formatter.date(from: startStr)
                ?? ISO8601DateFormatter().date(from: startStr) else {
            return []
        }

        let daysInPhase = Int(Date().timeIntervalSince(startDate) / 86400)
        var recs: [Recommendation] = []

        switch phase {
        case .vegetative:
            if plant.plantType == .photo, daysInPhase >= 42 {
                recs.append(Recommendation(
                    plantId: plant.id,
                    source: "rules",
                    message: "\(plant.name) has been in veg for \(daysInPhase) days \u{2014} ready to flip to flower?",
                    severity: "info"
                ))
            }
            if plant.plantType == .auto, daysInPhase >= 21 {
                recs.append(Recommendation(
                    plantId: plant.id,
                    source: "rules",
                    message: "\(plant.name) has been in veg for \(daysInPhase) days \u{2014} most autos start flowering around now",
                    severity: "info"
                ))
            }
        case .drying:
            if daysInPhase >= 7 {
                recs.append(Recommendation(
                    plantId: plant.id,
                    source: "rules",
                    message: "\(plant.name) has been drying for \(daysInPhase) days \u{2014} check if small stems snap cleanly",
                    severity: "info"
                ))
            }
        default:
            break
        }

        return recs
    }

    // MARK: - Evaluate

    /// Evaluate a plant against a sensor reading and return recommendations.
    /// Phase-aware: skips sensor checks for non-monitored phases and uses
    /// phase-specific thresholds when available.
    static func evaluatePlant(
        plant: Plant,
        reading: SensorReading
    ) -> [Recommendation] {
        // Skip all sensor checks for non-monitored phases
        if let phase = plant.currentPhase, !phase.hasMonitoring {
            return []
        }

        let thresholds = resolveThresholds(plant: plant)
        var recs: [Recommendation] = []

        // Moisture too low
        if let moisture = reading.moisture, moisture < thresholds.moistureMin {
            recs.append(Recommendation(
                plantId: plant.id,
                source: "rules",
                message: "\(plant.name) needs water! Soil moisture at \(moisture)% (minimum: \(thresholds.moistureMin)%)",
                severity: moisture < 20 ? "urgent" : "warning"
            ))
        }

        // Moisture too high
        if let moisture = reading.moisture, moisture > thresholds.moistureMax {
            recs.append(Recommendation(
                plantId: plant.id,
                source: "rules",
                message: "\(plant.name) may be overwatered \u{2014} moisture at \(moisture)% (maximum: \(thresholds.moistureMax)%)",
                severity: "warning"
            ))
        }

        // Temperature too low
        if let temp = reading.temperature, temp < thresholds.tempMin {
            recs.append(Recommendation(
                plantId: plant.id,
                source: "rules",
                message: "Too cold for \(plant.name)! Temperature at \(temp)\u{00B0}C (minimum: \(thresholds.tempMin)\u{00B0}C)",
                severity: temp < thresholds.tempMin - 5 ? "urgent" : "warning"
            ))
        }

        // Temperature too high
        if let temp = reading.temperature, temp > thresholds.tempMax {
            recs.append(Recommendation(
                plantId: plant.id,
                source: "rules",
                message: "Too hot for \(plant.name)! Temperature at \(temp)\u{00B0}C (maximum: \(thresholds.tempMax)\u{00B0}C)",
                severity: temp > thresholds.tempMax + 5 ? "urgent" : "warning"
            ))
        }

        // Low battery
        if let battery = reading.battery, battery < 15 {
            recs.append(Recommendation(
                plantId: plant.id,
                source: "rules",
                message: "Sensor battery low for \(plant.name) (\(battery)%) \u{2014} charge soon",
                severity: battery < 5 ? "urgent" : "warning"
            ))
        }

        // High-light plant in low light — only in growing phases or when phase is nil (backward compat)
        let applyLightRule = plant.currentPhase == nil || (plant.currentPhase?.isGrowing ?? false)
        if applyLightRule, let light = reading.light, plant.lightPreference == "high", light < 1000 {
            recs.append(Recommendation(
                plantId: plant.id,
                source: "rules",
                message: "\(plant.name) prefers bright light but only getting \(light) lux \u{2014} consider moving to a sunnier spot",
                severity: "info"
            ))
        }

        // Append transition suggestions
        recs.append(contentsOf: transitionSuggestions(plant: plant))

        return recs
    }

    /// Process a reading: evaluate rules and persist any recommendations to the database.
    static func processReading(plant: Plant, reading: SensorReading, db: AppDatabase) throws {
        let recs = evaluatePlant(plant: plant, reading: reading)
        for rec in recs {
            try db.addRecommendation(rec)
        }
    }
}
